::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: Copyright (c) 2018 Mirantis Inc., Enea AB and others.
:
: All rights reserved. This program and the accompanying materials
: are made available under the terms of the Apache License, Version 2.0
: which accompanies this distribution, and is available at
: http://www.apache.org/licenses/LICENSE-2.0
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Fri, 29 Jun 2018 02:48:55 +0200
Subject: [PATCH] maasng: Allow fabric ID to be determined from CIDR

MaaS usually creates network resources (e.g. subnets, fabrics)
automatically based on existing network interface definitions.
However, the fabric naming is not guaranteed across deployments,
so the same interface/subnet CIDR might end up as 'fabric-2' or
'fabric-5' depending on non-deterministic factors.

To allow configuring VLAN DHCP on such network segments we need a
mechanism that does not hardcode the fabric ID in pillar data, but
instead determines it at runtime based on subnet CIDR.

Fix 'name' arg in module.run calls

According to the docs, if a module function accepts an argument called,
name, then m_name must be used to specify that argument, to avoid a
collision with the name argument.

Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 README.rst         | 20 +++++++++++++++++++-
 _modules/maasng.py |  7 ++++++-
 maas/region.sls    | 34 ++++++++++++++++++++++++++--------
 3 files changed, 51 insertions(+), 10 deletions(-)

diff --git a/_modules/maasng.py b/_modules/maasng.py
index ce8e99e..706c0ec 100644
--- a/_modules/maasng.py
+++ b/_modules/maasng.py
@@ -966,7 +966,12 @@ def get_fabricid(fabric):
     try:
         return list_fabric()[fabric]['id']
     except KeyError:
-        return {"error": "Frabic not found on MaaS server"}
+        # fabric might be specified as CIDR, try to find the actual fabric ID
+        maas_subnets = list_subnet()
+        for subnet in maas_subnets.keys():
+            if maas_subnets[subnet]['cidr'] == fabric:
+                return maas_subnets[subnet]['vlan']['fabric_id']
+        return {"error": "Fabric not found on MaaS server"}


 def update_vlan(name, fabric, vid, description, primary_rack, dhcp_on=False):
diff --git a/maas/region.sls b/maas/region.sls
index f50b901..79a99fe 100644
--- a/maas/region.sls
+++ b/maas/region.sls
@@ -293,9 +293,10 @@ maas_fabrics:
 {%- if region.subnets is defined %}
 {%- for subnet_name, subnet in region.subnets.iteritems() %}
 maas_create_subnet_{{ subnet_name }}:
-  maasng.subnet_present:
+  module.run:
+  - name: maasng.subnet_present
   - cidr: {{ subnet.cidr }}
-  - name: {{ subnet_name }}
+  - m_name: {{ subnet_name }}
   - fabric: {{ subnet.fabric }}
   - gateway_ip: {{ subnet.gateway_ip }}
   - require:
@@ -305,12 +306,27 @@ maas_create_subnet_{{ subnet_name }}:
     {%- endif %}
 {%- endfor %}

+{%- for subnet_name, subnet in region.subnets.iteritems() %}
+{%- for vid, vlan in subnet.get('vlan', {}).items() %}
+maas_update_vlan_for_{{ subnet_name }}_{{ vid }}:
+  module.run:
+  - name: maasng.update_vlan
+  - vid: {{ vid }}
+  - fabric: {{ subnet.cidr }}
+  - m_name: {{ vlan.get('name', '') }}
+  - description: {{ vlan.description }}
+  - primary_rack: {{ region.maas_config.maas_name }}
+  - dhcp_on: {{ vlan.get('dhcp','False') }}
+{%- endfor %}
+{%- endfor %}
+
 {%- for subnet_name, subnet in region.subnets.iteritems() %}
 {%- if subnet.get('multiple') == True %}
 {%- for range_name, iprange in subnet.get('iprange',{}).items() %}
 maas_create_ipranger_{{ range_name }}:
-  maasng.iprange_present:
-  - name: {{ range_name }}
+  module.run:
+  - name: maasng.iprange_present
+  - m_name: {{ range_name }}
   - type_range: {{ iprange.type }}
   - start_ip: {{ iprange.start }}
   - end_ip: {{ iprange.end }}
@@ -320,8 +336,9 @@ maas_create_ipranger_{{ range_name }}:
 {%- endfor %}
 {%- else %}
 maas_create_ipranger_{{ subnet_name }}:
-  maasng.iprange_present:
-  - name: {{ subnet.get('cidr', []) }}
+  module.run:
+  - name: maasng.iprange_present
+  - m_name: {{ subnet.get('cidr', []) }}
   - type_range: {{ subnet.iprange.type }}
   - start_ip: {{ subnet.iprange.start }}
   - end_ip: {{ subnet.iprange.end }}
@@ -372,10 +389,11 @@ maas_domain:
 {%- for fabric_name, fabric in region.fabrics.iteritems() %}
 {%- for vid, vlan in fabric.get('vlan',{}).items() %}
 maas_update_vlan_for_{{ fabric_name }}_{{ vid }}:
-  maasng.update_vlan:
+  module.run:
+  - name: maasng.update_vlan
   - vid: {{ vid }}
   - fabric: {{ fabric_name }}
-  - name: {{ vlan.get('name','') }}
+  - m_name: {{ vlan.get('name','') }}
   - description: {{ vlan.description }}
   - primary_rack: {{ region.maas_config.maas_name }}
   - dhcp_on: {{ vlan.get('dhcp','False') }}
